kani-compiler/src/kani_middle/points_to/points_to_analysis.rs

```diff
@@ -202,161 +204,119 @@ impl<'a, 'tcx> Analysis<'tcx> for PointsToAnalysis<'a, 'tcx> {
             match instance.def {
                 // Intrinsics could introduce aliasing edges we care about, so need to handle them.
                 InstanceKind::Intrinsic(def_id) => {
-                    match self.tcx.intrinsic(def_id).unwrap().name.to_string().as_str() {
-                        name if name.starts_with("atomic") => {
-                            match name {
-                                // All `atomic_cxchg` intrinsics take `dst, old, src` as arguments.
-                                // This is equivalent to `destination = *dst; *dst = src`.
-                                name if name.starts_with("atomic_cxchg") => {
-                                    assert_eq!(
-                                        args.len(),
-                                        3,
-                                        "Unexpected number of arguments for `{name}`"
-                                    );
-                                    assert!(matches!(
-                                        args[0].node.ty(self.body, self.tcx).kind(),
-                                        TyKind::RawPtr(_, Mutability::Mut)
-                                    ));
-                                    let src_set =
-                                        self.successors_for_operand(state, args[2].node.clone());
-                                    let dst_set =
-                                        self.successors_for_deref(state, args[0].node.clone());
-                                    let destination_set =
-                                        state.resolve_place(*destination, self.instance);
-                                    state.extend(&destination_set, &state.successors(&dst_set));
-                                    state.extend(&dst_set, &src_set);
-                                }
-                                // All `atomic_load` intrinsics take `src` as an argument.
-                                // This is equivalent to `destination = *src`.
-                                name if name.starts_with("atomic_load") => {
-                                    assert_eq!(
```

Comment:
Maybe I'm reading it wrong, but where did these assertions about the length of arguments go?

And perhaps, now that we have a separate type for this, maybe we should have some method like `Intrinsic::arg_len() -> usize` or `Intrinsic::arg_len_matches(usize) -> bool` that checks if the argument number is correct in one place?

Reply:
These assertions were duplicating similar checks that happen in the codegen implicitly, so they seemed unnecessary and fragile to me. 

I agree that we need a better way to ensure argument types/length for the intrinsics. Perhaps we can make the `Intrinsic` enum constructor to accept a function signature (e.g. a list of StableMIR types) that is validated against what we expect it to be?